File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.java
class MedianFinder {
    public int size;
    public PriorityQueue<Integer> maxpq;
    public PriorityQueue<Integer> minpq;

    public MedianFinder() {
        size = 0;
        maxpq = new PriorityQueue<Integer>(Collections.reverseOrder());
        minpq = new PriorityQueue<Integer>();
    }
    
    public void addNum(int num) {
        if (size == 0)
        {
            maxpq.add(num);
            size++;
        }
        else
        {
            if (size%2 == 0)
            {
                if (num < minpq.peek())
                {
                    maxpq.add(num);
                }
                else
                {
                    int a = minpq.poll();
                    minpq.add(num);
                    maxpq.add(a);
                }
                size++;
            }
            else
            {
                if (num > maxpq.peek())
                {
                    minpq.add(num);
                }
                else
                {
                    int a = maxpq.poll();
                    maxpq.add(num);
                    minpq.add(a);
                }
                size++;
            }
        }
        
    }
    
    public double findMedian() {
        double c;
        if (size%2 == 0)
        {
            c = (double)(maxpq.peek() + minpq.peek())/2;
        }
        else
        {
            c = (double)maxpq.peek();
        }
        return c;
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
class MedianFinder(object):

    def __init__(self):
        self.leftheap = []
        self.rightheap = []

    def addNum(self, num):
        """
        :type num: int
        :rtype: None
        """
        if len(self.leftheap) == 0 or len(self.rightheap) == 0:
            heapq.heappush(self.leftheap, -num)
        # elif len(self.rightheap) == 0:
        #     heapq.heappush(self.rightheap, num)
        else:
            if num <= -self.leftheap[0]:
                heapq.heappush(self.leftheap, -num)
            else:
                heapq.heappush(self.rightheap, num)
        if len(self.leftheap) - len(self.rightheap) == 2:
                temp = -heapq.heappop(self.leftheap)
                heapq.heappush(self.rightheap, temp)
        if len(self.rightheap) - len(self.leftheap) == 2:
                temp = heapq.heappop(self.rightheap)
                heapq.heappush(self.leftheap, -temp)


    def findMedian(self):
        """
        :rtype: float
        """
        if len(self.leftheap) == len(self.rightheap):
            return (float)(-self.leftheap[0] + self.rightheap[0])/2
        elif len(self.leftheap) > len(self.rightheap):
            return -self.leftheap[0]
        else:
            return self.rightheap[0]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()